### ISSUE TITLE
mdList: Expando feature needed.

### ISSUE THREAD

**Issue by marcysutton** [angular/material#985]
A nice feature in the Material Design spec is the Expand/collapse [List-Item Control](http://www.google.com/design/spec/components/lists-controls.html#lists-controls-types-of-list-controls). 

![Material Design spec for expandable list control](https://cloud.githubusercontent.com/assets/1045233/5495051/fbd26f56-86ae-11e4-9f51-35365bfe51e7.png)

> As part of the Angular Material [Documentation Re-design](https://github.com/angular/material/issues/916), this feature will be used in the  `Sidenav`.
> This is related to mdList Issue #904.

```html
<md-list ng-repeat="topic in menu.activities">

   <md-list-item md-expando>
      <md-icon src="{{topic.icon}}"></md-icon>
     <label>{{topic.title}}</label>
   </md-list-item>

</md-list>  
```
<h2>Visual requirements:</h2>
 * Caret changes to indicate open/closed
 * Expanded panel has top and bottom borders
 * Child items are indented in line with parent toggle item

<h2>Functional requirements:</h2>
 * Hidden panels should be really hidden with `display: none` (or `aria-hidden="true"` and `tabIndex="-1"` on each child control)
 * Animated transitions

<h2>Semantic requirements:</h2>
 * Parent toggle item should indicate the child list it owns with `aria-controls="childListId"`
 * `aria-expanded="true | false"` on parent toggle item to indicate whether it is expanded or collapsed
 * Either `role="menu"` and `role="menuitem"` [should be used to indicate child elements are interactive](http://www.w3.org/TR/wai-aria/roles#menuitem) OR list items should have `<md-button>` inside of them ([related issue](https://github.com/angular/material/issues/904))

<h2>Questions:</h2>
 * What is the style/indentation for items that do not have visual icons?
 * What is the style for the hover on the caret control?

**Comment by shannonmpoole** [comment]
:+1:

**Comment by SweVictor** [comment]
I was looking for how to make a list item clickable, much like the [Google Material Inbox example](http://www.google.com/design/spec/components/lists.html). I came here through #904, hopefully this is the right place.

Anyway - after experimenting with the current release, I realized that wrapping a list item i an <a> tag looks very un-material-like and wrapping an item with a md-button completely ruins the layout.

I found the discussion in #904 slightly hard to follow, so you might be on to this already, anyway my request is that Google Material's "primary/secondary action" (linked page above => Content => Tile actions) is implemented.

Most simple case would be the list example from the docs, but clickable - more advanced would be a secondary action for the image and primary for the rest. Anyway, a large click target would one goal for me.

A simple implementation to maybe look at would be [Bootstrap's clickable list items](http://getbootstrap.com/components/#list-group-linked), a control we use heavily on our Bootstrap projects.

/Victor

**Comment by mikol** [comment]
@SweVictor I have a topic open in the forum to discuss just the functionality that you are describing. Please [take a look](https://groups.google.com/d/msgid/ngmaterial/6f5df85b-9813-4d85-a303-8030bf4e2387%40googlegroups.com).

**Comment by marcysutton** [comment]
Closed with abb807d0dcb77b92132b5ab73d61f97aa953f461. This feature will be available in 0.9.0!

**Comment by marcysutton** [comment]
Actually that feature didn't include expand/collapse yet, so I have reopened this issue. Sorry!

**Comment by anupbelambe** [comment]
I hope we are supporting the keyboard navigation for the same. If not, appreciate if the same can be added.
e.g. http://demos.telerik.com/kendo-ui/treeview/index

**Comment by robob4him** [comment]
@marcysutton Google has updated the docs. Do you mean: http://www.google.com/design/spec/components/lists-controls.html#lists-controls-types-of-list-controls

**Comment by marcysutton** [comment]
Yep! I have updated the link in the issue description, it is the same content with a new URL. Thanks!

**Comment by ThomasBurleson** [comment]
Expando lists will be implemented in 0.10.

**Comment by anupbelambe** [comment]
Following is the ARIA guideline for keyboard navigation for list.
http://www.w3.org/WAI/PF/aria-practices/#Listbox

**Comment by marcysutton** [comment]
Listbox is a different role. The exact semantics are TBD.

**Comment by anupbelambe** [comment]
My bad. But I hope the keyboard interaction will be ARIA spec compliant and will be fixed as part of this issue.

**Comment by marcysutton** [comment]
That's the idea! :)

**Comment by anupbelambe** [comment]
Awesome Thanks.

**Comment by brianfeister** [comment]
Curious of this is still slated for 0.9.0?  Any change there?  Thanks for accommodating this feature, I've already run into the need for this twice in just a few weeks of using Angular Material.

**Comment by realtebo** [comment]
Is this beautiful feature still planned? Or, at least, can you provide a tutorial, or  a demo, on how to realize this waiting feature until the feature itself is not developed?

Thanks for all your work. Angular Material is amazing.

**Comment by dreamsoftech** [comment]
:+1:

**Comment by dparish** [comment]
+1 this really would be a nice feature to have.

**Comment by jamm** [comment]
please don't post "+1" spam

**Comment by dparish** [comment]
I don't mean it as spam.  Much like the earlier thumbs up, I'm just trying to vote for the feature.  I get that it clutters the discussion though. Is there another way to "vote" for the feature to make sure it's up in the priority list?

**Comment by ThomasBurleson** [comment]
@dparish - we really don't have a better way of voting... (wish list for GitHub). So thanks for your vote.
@jamm - While I am not a fan of +1, but for voting/endorsement  on Github it works.

**Comment by ThomasBurleson** [comment]
We are currently hyper-focused on  a summer release for Angular Material v1.0; which will include Menu and DateTimePicker components, performance enhancements, bug fixes, and improved documentation.

The Expando and Accordion features are definitely on our roadmap but not critical to a v1.0 Beta candidate. Time-permitting, a demo of this feature will be posted on CodePen. 

> We are locking this issue for now... it will remain an **open** issue, but locked to future discussions. Please future discussions in the [Angular Material Forum](https://groups.google.com/forum/#!forum/ngmaterial)

**Comment by ThomasBurleson** [comment]
This issue is closed as part of our ‘Surge Focus on Material 2' efforts.
        For details, see our [forum posting](http://bit.ly/1UhZyWs).

**Comment by ThomasBurleson** [comment]
This issue is closed as part of our ‘Surge Focus on Material 2' efforts.
      For details, see our forum posting @ http://bit.ly/1UhZyWs.
